File: archive/python/src/index/keyword_index.py
import re
from rank_bm25 import BM25Okapi
from src.models import Chunk
from src.notebook.store import NotebookStore
WORD = re.compile(r"[a-z0-9_]+")

def tokenize(text: str) -> list[str]:
    return WORD.findall(text.lower())

class KeywordIndex:
    def __init__(self, chunk_ids: list[str], corpus: list[list[str]]):
        self.chunk_ids = chunk_ids
        self.bm25 = BM25Okapi(corpus) if corpus else None
        
    @classmethod
    def for_notebook(
        cls, notebook: str, store: NotebookStore | None = None
    ) -> "KeywordIndex":
        chunks: list[Chunk] = (store or NotebookStore()).child_chunks(notebook)
        return cls(
            [chunk.chunk_id for chunk in chunks],
            [tokenize(chunk.text) for chunk in chunks],
        )

    def search(self, query: str, k: int) -> list[tuple[str, float]]:
        """Returns [(chunk_id, bm25_score)], best first, zero-scores dropped."""
        if self.bm25 is None:
            return []
        scores = self.bm25.get_scores(tokenize(query))
        ranked = sorted(zip(self.chunk_ids, scores), key = lambda pair: pair[1],
                        reverse = True)
        return [(chunk_id, float(score)) for chunk_id, score in ranked[:k] if score > 0] 

File: src/search/keywordIndex.ts
import { BM25 } from "./bm25.ts";
import { NotebookStore } from "../notebook/store.ts";

const WORD = /[a-z0-9_]+/g;

export function tokenize(text: string): string[] {
  return text.toLowerCase().match(WORD) ?? [];
}

export class KeywordIndex {
  private chunkIds: string[];
  private documentIds: string[];
  private bm25: BM25 | null;

  constructor(chunkIds: string[], corpus: string[][], documentIds: string[] = []) {
    this.chunkIds = chunkIds;
    this.documentIds = documentIds;
    this.bm25 = corpus.length > 0 ? new BM25(corpus) : null;
  }

  static forNotebook(notebook: string, store: NotebookStore = new NotebookStore()): KeywordIndex {
    const chunks = store.childChunks(notebook);
    return new KeywordIndex(
      chunks.map((chunk) => chunk.chunkId),
      chunks.map((chunk) => tokenize(chunk.text)),
      chunks.map((chunk) => chunk.documentId),
    );
  }

  search(query: string, k: number, documentIds?: readonly string[]): [string, number][] {
    if (!this.bm25) return [];
    const scope = documentIds && documentIds.length > 0 ? new Set(documentIds) : null;
    const scored: [string, number][] = [];
    for (const [doc, score] of this.bm25.scores(tokenize(query))) {
      if (score <= 0) continue;
      const chunkId = this.chunkIds[doc]!;
      if (scope && !scope.has(this.documentIds[doc] ?? "")) continue;
      scored.push([chunkId, score]);
    }
    scored.sort((a, b) => b[1] - a[1]);
    return scored.slice(0, k);
  }

  get size(): number {
    return this.chunkIds.length;
  }
}
